File: healthcheck_localization.py
#!/usr/bin/python3
import rclpy
import os
from geometry_msgs.msg import PoseWithCovarianceStamped
from rclpy.qos import QoSProfile, QoSDurabilityPolicy

print ("SLAM_MODE:" + os.environ.get("SLAM_MODE"))

def callback(msg):
    print(f"received amcl_pose")
    exit(0)

rclpy.init(args=[])
latching_qos = QoSProfile(depth=1,
            durability=QoSDurabilityPolicy.TRANSIENT_LOCAL)

node = rclpy.create_node("amcl_pose_listener")
sub = node.create_subscription(PoseWithCovarianceStamped, "/amcl_pose", callback, qos_profile=latching_qos)
rclpy.spin_once(node)
print(f"amcl_pose timeout")
exit(1)

File: healthcheck_localization.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/pose_with_covariance_stamped.hpp"
#include "cstdlib"

using namespace std::chrono_literals;

#define TIMEOUT 2s

int msg_received = EXIT_FAILURE;

void msg_callback(const geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr msg)
{
  std::cout << "Message received" << std::endl;
  msg_received = EXIT_SUCCESS;
  rclcpp::shutdown();
}

void timeout_callback()
{
  std::cout << "Timeout" << std::endl;
  rclcpp::shutdown();
}

int main(int argc, char* argv[])
{
  rclcpp::init(argc, argv);
  auto node = rclcpp::Node::make_shared("healthcheck_node");
  rclcpp::QoS qos(1);
  qos.transient_local();
  auto sub = node->create_subscription<geometry_msgs::msg::PoseWithCovarianceStamped>("/amcl_pose", qos, msg_callback);
  auto timer = node->create_wall_timer(TIMEOUT, timeout_callback);

  rclcpp::spin(node);
  return msg_received;
}

